// carbon-components-angular/esm2020/treeview/treeview.service.mjs
import { Injectable } from "@angular/core";
import { ReplaySubject } from "rxjs";
import * as i0 from "@angular/core";
export class TreeViewService {
    constructor() {
        /**
         * Variable used across all nodes and wrapper to determine if we should allow content projection
         * or generate the tree
         *
         * Value is updated by passing a value to `tree` input in wrapper component.
         */
        this.contentProjected = true;
        /**
         * **Experimental**
         */
        this.isMultiSelect = false;
        this.selectionSubject = new ReplaySubject(1);
        /**
         * Hold's list of selected nodes and preserves order
         */
        this.value = new Map();
        this.selectionObservable = this.selectionSubject.asObservable();
    }
    /**
     * Store selected node in map
     * @param node: Node
     */
    selectNode(node) {
        if (!node) {
            return;
        }
        // Since multiselect is not enabled, we clear existing map
        if (!this.isMultiSelect) {
            this.value.clear();
        }
        this.value.set(node.id, node);
        this.selectionSubject.next(this.value);
    }
}
TreeViewService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeViewService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
TreeViewService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeViewService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeViewService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
